src/listener/tcp_listener.rs

```diff
@@ -69,11 +62,41 @@ fn handle_tcp<State: Clone + Send + Sync + 'static>(app: Server<State>, stream:
 }
 
 #[async_trait::async_trait]
-impl<State: Clone + Send + Sync + 'static> Listener<State> for TcpListener {
-    async fn listen(&mut self, app: Server<State>) -> io::Result<()> {
-        self.connect().await?;
-        let listener = self.listener()?;
-        crate::log::info!("Server listening on {}", self);
+impl<State> Listener<State> for TcpListener<State>
+where
+    State: Clone + Send + Sync + 'static,
+{
+    async fn bind(&mut self, server: Server<State>) -> io::Result<()> {
+        assert!(self.server.is_none(), "`bind` should only be called once");
```

Comment:
An Err might be nicer here.

Reply:
Same as above: failing early is very much the intended behavior. Perhaps we can encode it in the type system in the future, but for now this is intended to fail before booting and signal the programmer needs to change their usage patterns -- rather than signaling an error that originates from the the operating system (e.g. "port already in use" or similar).